File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def bfs(x, y, maps, n, m):
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]
    q = deque()
    q.append((x, y))
    
    while q:
        x, y = q.popleft()
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < n and 0 <= ny < m:
                if maps[nx][ny] == 1:
                    maps[nx][ny] = maps[x][y] + 1
                    q.append((nx, ny))

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    bfs(0, 0, maps, n, m)
    if maps[n - 1][m - 1] == 1:
        return -1
    else:
        return maps[n - 1][m - 1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const dy = [1, -1, 0, 0];
    const dx = [0, 0, 1, -1];
    const row = maps.length;
    const col = maps[0].length;
    
    const visitCount = [...maps].map((r) => r.map((c) => 1));
    const q = [[0, 0]];
    
    while(q.length) {
        const [y, x] = q.shift();
        
        for(let i = 0; i < 4; i++) {
            const ny = y + dy[i];
            const nx = x + dx[i];
            
            if (ny >= 0 && nx >= 0 && ny < row && nx < col) {
                if (maps[ny][nx] === 1 && visitCount[ny][nx] === 1) {
                    visitCount[ny][nx] = visitCount[y][x] + 1;
                    q.push([ny, nx]);
                }
            }
        }
    }
    
    return visitCount[row-1][col-1] === 1 ? -1 : visitCount[row-1][col-1];
}